### PR TITLE
Add chain ID attribute

### PR DESCRIPTION
**Pull request by lameiraatt**
This OTEP proposes the addition of a chain ID attribute to spans and logs that supports data processing/analysis and can be implemented as part of an extension to the default OTel SDK.

Feedback welcome and appreciated! 😊

### CODE REVIEW

File: text/0228-chain-id.md [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+
+The example span above includes a `task.processing.time.ns` attribute, which represents the amount of time the CPU spent processing the task associated with that span.
+If we want to determine the total amount of CPU time spent on all tasks performed under that span (level 3+), we can:
+
+1. Find spans where the `chain.id` attribute value starts with `2e072d7d02464a2490b65c864da59609#5#31#739#11`, to get the entire subtree;
+2. Extract `task.processing.time.ns` values, if the attribute is present;
+3. Sum.
+
+Without chain ID, we'd have to first create the tree, using span ID and parent span ID information.
+When processing big data sets, with millions of data points, generated by thousands of processes, this can have a considerable performance impact.
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply text/0228-chain-id.md:57]: "millions of data points" is not necessarily a Google-scale problem. I've seen smaller companies processing tracing volume comparable to hyper-scalers, the difference is that hyper-scalers just have to sample more.

And to me, the challenge with processing mentioned here is not necessarily with the data volume, but with expressiveness of the query languages. Aside from TempoDB QL I am not aware of any QL for traces that allows to express queries against the graph (you can do one-level parent/child in SQL, anything more becomes too cumbersome to express). The chain-id largely solves this issue by allowing to query into tabular data set without any aggregation of events into traces.

[the remark this answers is not in the document]
Comment (electron0zero) [reply text/0228-chain-id.md:57]: Thank you Yuri, Here is [TraceQL Design Proposal](https://github.com/grafana/tempo/blob/main/docs/design-proposals/2022-04%20TraceQL%20Concepts.md) if you want to read more on it.

**Comment by bputt-e** [comment]
Does this proposal help with linked spans? Let's say I have 500 traces that get 'merged' into a new trace....What would that new trace look like, would it have any 'historical' knowledge to add any chain info or would it be considered a new trace and have nothing starting out.

**Comment by joe-elliott** [comment]
fwiw, Tempo achieves similar functionality by computing the [nested set model](https://en.wikipedia.org/wiki/Nested_set_model) for the tree upon storage. if this were adopted by OTEL it would make calculating the nested set quite a bit easier.

the UUID passed seems to be used to track the root process instance. imo it would be better to make this separate from the concept of the hierarchy. perhaps two different attributes that could be independently controlled:

Perhaps something like:
`chain.hierarchy`
`chain.root`

i'm also a bit concerned about traces with extreme depth creating enormous chain ids. perhaps a configuration option for max depth recorded would be worthwhile?

**Comment by tedsuo** [comment]
@lameiraatt  we are cleaning up stale OTEP PRs. If there is no further action at this time, we will close this PR in one week. Feel free to open it again when it is time to pick it back up.

**Comment by carlosalberto** [comment]
OTEPs have been moved to the [Specification](https://github.com/open-telemetry/opentelemetry-specification/tree/main/oteps/) repository. Please consider re-opening this PR against the new location. Closing.

**Pull request closed without merging**